### ISSUE TITLE
Docs/Nx Integration Guide

### ISSUE THREAD

**Issue by thanos** [thanos/ExZarr#183]
**Purpose:**
Show how ExZarr integrates with Nx (Numerical Elixir) for numerical computing workflows. Cover persisting Nx tensors, reading arrays into Nx, dtype/shape mapping, and practical patterns for ML/scientific computing. Position ExZarr as persistent storage layer for Nx computations.

**Required Subsections:**
1. Nx and ExZarr Overview
2. Persisting Nx Tensors to ExZarr
3. Reading ExZarr Arrays into Nx Tensors
4. Dtype and Shape Mapping
5. Batch Processing Patterns
6. ML Training/Inference Workflows
7. Performance Considerations
8. Limitations and Workarounds

**Content Requirements:**

*Nx and ExZarr Overview:*
- Explain the relationship:
  - **Nx**: In-memory numerical computing (like NumPy)
  - **ExZarr**: Persistent, chunked array storage (like HDF5)
  - Together: Complete numerical computing stack for Elixir

- Use cases:
  - Persist Nx computation results for later analysis
  - Load large datasets that don't fit in memory
  - Share numerical data between Nx-based applications
  - Checkpoint ML model weights and training data
  - Archive scientific computation outputs

- Conceptual mapping:
  ```
  Nx.Tensor (in-memory) ←→ ExZarr.Array (persistent, chunked)
  Nx.shape()            ←→ array.metadata.shape
  Nx.type()             ←→ array.metadata.dtype
  Nx operations         ←→ ExZarr.Array.get_slice/set_slice
  ```

*Persisting Nx Tensors to ExZarr:*
- Basic example (full tensor):
  ```elixir
  # Create Nx tensor
  tensor = Nx.iota({1000, 1000}, type: :f64)

  # Create ExZarr array with matching shape/dtype
  {:ok, array} = ExZarr.create(
    shape: Nx.shape(tensor),
    chunks: {100, 100},
    dtype: nx_to_zarr_dtype(Nx.type(tensor)),
    storage: :filesystem,
    path: "/data/my_tensor"
  )

  # Convert tensor to nested tuples
  data = Nx.to_list(tensor)
        |> nested_list_to_tuple()

  # Write to array
  :ok = ExZarr.Array.set_slice(array, data,
    start: {0, 0},
    stop: Nx.shape(tensor)
  )
  ```

- Helper function for dtype conversion:
  ```elixir
  defmodule ExZarr.Nx.Helpers do
    def nx_to_zarr_dtype({:s, 8}), do: :int8
    def nx_to_zarr_dtype({:s, 16}), do: :int16
    def nx_to_zarr_dtype({:s, 32}), do: :int32
    def nx_to_zarr_dtype({:s, 64}), do: :int64
    def nx_to_zarr_dtype({:u, 8}), do: :uint8
    def nx_to_zarr_dtype({:u, 16}), do: :uint16
    def nx_to_zarr_dtype({:u, 32}), do: :uint32
    def nx_to_zarr_dtype({:u, 64}), do: :uint64
    def nx_to_zarr_dtype({:f, 32}), do: :float32
    def nx_to_zarr_dtype({:f, 64}), do: :float64

    def zarr_to_nx_dtype(:int8), do: {:s, 8}
    def zarr_to_nx_dtype(:int16), do: {:s, 16}
    def zarr_to_nx_dtype(:int32), do: {:s, 32}
    def zarr_to_nx_dtype(:int64), do: {:s, 64}
    def zarr_to_nx_dtype(:uint8), do: {:u, 8}
    def zarr_to_nx_dtype(:uint16), do: {:u, 16}
    def zarr_to_nx_dtype(:uint32), do: {:u, 32}
    def zarr_to_nx_dtype(:uint64), do: {:u, 64}
    def zarr_to_nx_dtype(:float32), do: {:f, 32}
    def zarr_to_nx_dtype(:float64), do: {:f, 64}

    def nested_list_to_tuple(list) when is_list(list) do
      list
      |> Enum.map(&nested_list_to_tuple/1)
      |> List.to_tuple()
    end
    def nested_list_to_tuple(value), do: value
  end
  ```

- Chunked write pattern (large tensors):
  ```elixir
  # For tensors larger than memory, write in chunks
  defmodule ExZarr.Nx.ChunkedWrite do
    def write_tensor_chunked(tensor, array, chunk_size) do
      {height, width} = Nx.shape(tensor)
      {chunk_h, chunk_w} = chunk_size

      for i <- 0..div(height - 1, chunk_h),
          j <- 0..div(width - 1, chunk_w) do

        # Extract chunk from tensor
        start_h = i * chunk_h
        start_w = j * chunk_w
        end_h = min(start_h + chunk_h, height)
        end_w = min(start_w + chunk_w, width)

        chunk = Nx.slice(tensor, [start_h, start_w], [end_h - start_h, end_w - start_w])

        # Convert and write
        data = chunk |> Nx.to_list() |> nested_list_to_tuple()
        :ok = ExZarr.Array.set_slice(array, data,
          start: {start_h, start_w},
          stop: {end_h, end_w}
        )
      end

      :ok
    end
  end
  ```

*Reading ExZarr Arrays into Nx Tensors:*
- Basic example (full array):
  ```elixir
  # Open ExZarr array
  {:ok, array} = ExZarr.open(path: "/data/my_tensor")

  # Read entire array
  {:ok, data} = ExZarr.Array.get_slice(array,
    start: {0, 0},
    stop: array.metadata.shape
  )

  # Convert to Nx tensor
  tensor = data
          |> tuple_to_nested_list()
          |> Nx.tensor(type: zarr_to_nx_dtype(array.metadata.dtype))
  ```

- Slice reading (memory-efficient):
  ```elixir
  # Read only a portion of array
  {:ok, data} = ExZarr.Array.get_slice(array,
    start: {0, 0},
    stop: {100, 100}
  )

  # Convert to Nx tensor
  slice_tensor = data
                |> tuple_to_nested_list()
                |> Nx.tensor(type: {:f, 64})

  # Now perform Nx operations on slice
  result = slice_tensor
          |> Nx.add(10)
          |> Nx.multiply(2)
  ```

- Streaming pattern (process chunks sequentially):
  ```elixir
  defmodule ExZarr.Nx.StreamProcessor do
    def process_array_chunked(array, chunk_size, processor_fn) do
      {height, width} = array.metadata.shape
      {chunk_h, chunk_w} = chunk_size

      results = for i <- 0..div(height - 1, chunk_h),
                    j <- 0..div(width - 1, chunk_w) do

        # Read chunk
        {:ok, data} = ExZarr.Array.get_slice(array,
          start: {i * chunk_h, j * chunk_w},
          stop: {min((i + 1) * chunk_h, height), min((j + 1) * chunk_w, width)}
        )

        # Convert to tensor
        tensor = data |> tuple_to_nested_list() |> Nx.tensor()

        # Process with user function
        processor_fn.(tensor, {i, j})
      end

      {:ok, results}
    end
  end

  # Usage
  {:ok, stats} = ExZarr.Nx.StreamProcessor.process_array_chunked(
    array,
    {100, 100},
    fn tensor, _coords ->
      %{
        mean: Nx.mean(tensor) |> Nx.to_number(),
        max: Nx.reduce_max(tensor) |> Nx.to_number()
      }
    end
  )
  ```

*Dtype and Shape Mapping:*
- Complete compatibility table:

  | Nx Type | ExZarr Dtype | Compatible | Notes |
  |---------|--------------|------------|-------|
  | `{:s, 8}` | `:int8` | Yes | 8-bit signed |
  | `{:s, 16}` | `:int16` | Yes | 16-bit signed |
  | `{:s, 32}` | `:int32` | Yes | 32-bit signed |
  | `{:s, 64}` | `:int64` | Yes | 64-bit signed |
  | `{:u, 8}` | `:uint8` | Yes | 8-bit unsigned |
  | `{:u, 16}` | `:uint16` | Yes | 16-bit unsigned |
  | `{:u, 32}` | `:uint32` | Yes | 32-bit unsigned |
  | `{:u, 64}` | `:uint64` | Yes | 64-bit unsigned |
  | `{:f, 32}` | `:float32` | Yes | 32-bit float |
  | `{:f, 64}` | `:float64` | Yes | 64-bit float |
  | `{:bf, 16}` | N/A | No | BF16 not in Zarr spec |
  | `{:f, 16}` | N/A | No | FP16 not in Zarr spec |
  | `{:c, 64}` | N/A | No | Complex not supported |

- Shape conversion:
  - Nx: `{1000, 500, 3}` (tuple)
  - ExZarr: `{1000, 500, 3}` (tuple, same representation)
  - Direct compatibility (no conversion needed)

*Batch Processing Patterns:*
- Pattern 1: Map-reduce over chunks
  ```elixir
  defmodule ExZarr.Nx.MapReduce do
    def map_reduce(array, chunk_size, map_fn, reduce_fn, initial) do
      {height, width} = array.metadata.shape
      {chunk_h, chunk_w} = chunk_size

      chunk_coords = for i <- 0..div(height - 1, chunk_h),
                         j <- 0..div(width - 1, chunk_w),
                         do: {i, j}

      chunk_coords
      |> Task.async_stream(fn {i, j} ->
        # Read chunk
        {:ok, data} = ExZarr.Array.get_slice(array,
          start: {i * chunk_h, j * chunk_w},
          stop: {min((i + 1) * chunk_h, height), min((j + 1) * chunk_w, width)}
        )

        # Convert and map
        tensor = data |> tuple_to_nested_list() |> Nx.tensor()
        map_fn.(tensor)
      end, max_concurrency: 8)
      |> Enum.reduce(initial, fn {:ok, mapped}, acc ->
        reduce_fn.(mapped, acc)
      end)
    end
  end

  # Example: Compute global statistics
  stats = ExZarr.Nx.MapReduce.map_reduce(
    array,
    {100, 100},
    fn tensor -> %{sum: Nx.sum(tensor), count: Nx.size(tensor)} end,
    fn %{sum: s, count: c}, %{total_sum: ts, total_count: tc} ->
      %{total_sum: Nx.add(ts, s), total_count: tc + c}
    end,
    %{total_sum: Nx.tensor(0.0), total_count: 0}
  )

  mean = Nx.divide(stats.total_sum, stats.total_count) |> Nx.to_number()
  ```

*ML Training/Inference Workflows:*
- Pattern: Checkpoint model weights
  ```elixir
  defmodule MLCheckpoint do
    def save_model_weights(model_state, path) do
      # Assuming model_state is a map of layer_name => Nx.Tensor
      Enum.each(model_state, fn {layer_name, tensor} ->
        layer_path = Path.join(path, to_string(layer_name))

        {:ok, array} = ExZarr.create(
          shape: Nx.shape(tensor),
          chunks: chunk_shape_for_tensor(Nx.shape(tensor)),
          dtype: nx_to_zarr_dtype(Nx.type(tensor)),
          storage: :filesystem,
          path: layer_path,
          compressor: :zstd
        )

        data = tensor |> Nx.to_list() |> nested_list_to_tuple()
        :ok = ExZarr.Array.set_slice(array, data,
          start: List.duplicate(0, tuple_size(Nx.shape(tensor))),
          stop: Nx.shape(tensor)
        )
      end)

      :ok
    end

    def load_model_weights(path) do
      path
      |> File.ls!()
      |> Enum.map(fn layer_name ->
        layer_path = Path.join(path, layer_name)
        {:ok, array} = ExZarr.open(path: layer_path)
        {:ok, data} = ExZarr.Array.get_slice(array,
          start: List.duplicate(0, tuple_size(array.metadata.shape)),
          stop: array.metadata.shape
        )

        tensor = data |> tuple_to_nested_list() |> Nx.tensor(
          type: zarr_to_nx_dtype(array.metadata.dtype)
        )

        {String.to_atom(layer_name), tensor}
      end)
      |> Map.new()
    end
  end
  ```

- Pattern: Mini-batch data loading
  ```elixir
  defmodule DataLoader do
    def create_batch_iterator(array, batch_size) do
      {num_samples, _features} = array.metadata.shape
      num_batches = div(num_samples, batch_size)

      Stream.map(0..(num_batches - 1), fn batch_idx ->
        start_idx = batch_idx * batch_size
        end_idx = min(start_idx + batch_size, num_samples)

        {:ok, batch_data} = ExZarr.Array.get_slice(array,
          start: {start_idx, 0},
          stop: {end_idx, tuple_size(elem(array.metadata.shape, 1))}
        )

        batch_data
        |> tuple_to_nested_list()
        |> Nx.tensor(type: zarr_to_nx_dtype(array.metadata.dtype))
      end)
    end
  end

  # Usage in training loop
  batches = DataLoader.create_batch_iterator(training_data, batch_size: 32)

  Enum.reduce(batches, model_state, fn batch, state ->
    # Training step with Nx/Axon
    {loss, updated_state} = train_step(state, batch)
    updated_state
  end)
  ```

*Performance Considerations:*
- Conversion overhead:
  - Nested tuple ↔ Nx tensor conversion has cost
  - For large arrays, conversion dominates if not careful
  - Minimize conversions by working in chunks

- Memory efficiency:
  - Nx tensors live in process heap (different from ExZarr binaries)
  - Converting large arrays can spike memory usage
  - Use streaming/chunked patterns for arrays > 100 MB

- Optimization tips:
  - Align ExZarr chunks with Nx processing chunks
  - Pre-allocate Nx tensors when possible
  - Use Nx backends (EXLA) for GPU acceleration of processing
  - Batch multiple chunks before Nx operations

*Limitations and Workarounds:*
- **Limitation**: No zero-copy conversion
  - Zarr binary chunks → nested tuples → Nx tensor (2 copies)
  - Workaround: Process chunks, not full arrays

- **Limitation**: Nx BF16/FP16 types not in Zarr
  - Workaround: Store as FP32, quantize in memory with Nx

- **Limitation**: Complex numbers not supported
  - Workaround: Store real/imaginary as separate arrays

- **Limitation**: Ragged/variable-length tensors
  - Zarr requires fixed-shape arrays
  - Workaround: Pad to maximum size, store mask separately

**Style:**
- Practical, code-heavy
- Show complete working examples
- Emphasize memory efficiency
- Provide reusable helper modules

**Acceptance Criteria:**
- [ ] Complete Nx → ExZarr example (persist tensor)
- [ ] Complete ExZarr → Nx example (load tensor)
- [ ] Dtype/shape compatibility table
- [ ] Chunked processing patterns
- [ ] ML checkpoint/loading example
- [ ] Batch data loading example
- [ ] Performance optimization tips
- [ ] Clear limitations with workarounds

**Notes / Gotchas:**
- Conversion overhead is real (measure before optimizing)
- Nx tensors and ExZarr arrays have different memory models
- BF16/FP16 widely used in ML but not in Zarr spec
- ExZarr chunks and ML batch sizes should align when possible
- For very large datasets, consider chunked/streaming patterns
- Nx.to_list() can be slow for large tensors
- Consider using Nx defn compilation for processing pipelines
